### PR TITLE
Fix read back of XDP mode when using "driver" mode

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

We write xdpdrv but ip link returns "xdp" when we read it back.

think the bug was exposed by kernels being updated so that now xdpdrv mode is supported on veths.

## Todos

- [x] Tests
- [ ] Documentation
- [x] Release note

## Release Note

```release-note
Fix application of the iptables-mode XDP DoS protection program on devices that support driver mode.
```

### CODE REVIEW

File: felix/bpf/bpf.go [hunk 1]
```diff
@@ -1229,13 +1229,15 @@ func (b *BPFLib) GetXDPMode(ifName string) (XDPMode, error) {
...
-		XDPGeneric.String(): XDPGeneric,
-	}
-	for i := range s {
-		if mode, ok := allModes[s[i]]; ok {
-			return mode, nil
+		"xdp":        XDPDriver, // We write "xdpdrv" but read back "xdp"
+		"xdpoffload": XDPOffload,
+		"xdpgeneric": XDPGeneric,
+	}
+	for _, mode := range []string{"xdpgeneric", "xdpoffload", "xdp"} {
```
Comment (neiljerram) [inline felix/bpf/bpf.go:1236]: How about `for modeString, mode := range allModes { ... }` and `if f == modeString { return mode, nil }` ?  That would avoid repeating the strings that we're looking for?

File: felix/dataplane/linux/xdp_state.go [hunk 2]
```diff
@@ -468,10 +468,27 @@ func (s *xdpIPState) newXDPResyncState(bpfLib bpf.BPFDataplane, ipsSource ipsets
...
 		tag, tagErr := bpfLib.GetXDPTag(iface)
 		mode, modeErr := bpfLib.GetXDPMode(iface)
-		// error can happen when the program was not pinned in the bpf filesystem, so we say it's bogus anyway
-		bogus := tagErr != nil || tag != programTag || modeErr != nil || !isValidMode(mode, xpdModes)
+
+		var bogosityReasons []string
+		if tagErr != nil {
+			bogosityReasons = append(bogosityReasons, fmt.Sprintf("error getting tag: %s", tagErr.Error()))
+		} else if tag != programTag {
+			bogosityReasons = append(bogosityReasons, "loaded program doesn't match expected tag")
```
Comment (neiljerram) [inline felix/dataplane/linux/xdp_state.go:476]: log the mismatching tags here?

File: felix/dataplane/linux/xdp_state.go [hunk 3]
```diff
@@ -468,10 +468,27 @@ func (s *xdpIPState) newXDPResyncState(bpfLib bpf.BPFDataplane, ipsSource ipsets
...
+		}
+		if modeErr != nil {
+			bogosityReasons = append(bogosityReasons, fmt.Sprintf("error getting mode: %s", modeErr.Error()))
+		} else if !isValidMode(mode, xpdModes) {
```
Comment (neiljerram) [inline felix/dataplane/linux/xdp_state.go:480]: Probably worth correcting "xpd" while you're in the area.

**Review by neiljerram — APPROVED** [review]
LGTM but I've suggested some minor tweaks.

File: felix/bpf/bpf.go (hunk 1 above)
Comment (fasaxc) [reply felix/bpf/bpf.go:1236]: That was to ensure deterministic ordering in case both strings showed up in the output.  Will rework to make that clear.

**Pull request merged**